File: problem/regexredux/solution/python3/1.py
# The Computer Language Benchmarks Game
# https://salsa.debian.org/benchmarksgame-team/benchmarksgame/
#
# regex-dna program contributed by Dominique Wahli
# 2to3
# mp by Ahmad Syukri
# modified by Justin Peel
# converted from regex-dna program

from sys import stdin
from re import sub, findall
from multiprocessing import Pool

def init(arg):
    global seq
    seq = arg

def var_find(f):
    return len(findall(f, seq))

def main():
    seq = stdin.read()
    ilen = len(seq)

    seq = sub(ʼ>.*\n|\nʼ, ʼʼ, seq)
    clen = len(seq)

    pool = Pool(initializer = init, initargs = (seq,))

    variants = (
          ʼagggtaaa|tttaccctʼ,
          ʼ[cgt]gggtaaa|tttaccc[acg]ʼ,
          ʼa[act]ggtaaa|tttacc[agt]tʼ,
          ʼag[act]gtaaa|tttac[agt]ctʼ,
          ʼagg[act]taaa|ttta[agt]cctʼ,
          ʼaggg[acg]aaa|ttt[cgt]ccctʼ,
          ʼagggt[cgt]aa|tt[acg]accctʼ,
          ʼagggta[cgt]a|t[acg]taccctʼ,
          ʼagggtaa[cgt]|[acg]ttaccctʼ)
    for f in zip(variants, pool.imap(var_find, variants)):
        print(f[0], f[1])

    subst = {
          ʼtHa[Nt]ʼ : ʼ<4>ʼ, ʼaND|caN|Ha[DS]|WaSʼ : ʼ<3>ʼ, ʼa[NSt]|BYʼ : ʼ<2>ʼ,
          ʼ<[^>]*>ʼ : ʼ|ʼ, ʼ\\|[^|][^|]*\\|ʼ : ʼ-ʼ}
    for f, r in list(subst.items()):
        seq = sub(f, r, seq)

    print()
    print(ilen)
    print(clen)
    print(len(seq))

if __name__=="__main__":
    main()

File: problem/regexredux/solution/perl/1.pl
# The Computer Language Benchmarks Game
# https://salsa.debian.org/benchmarksgame-team/benchmarksgame/
#
# contributed by A. Sinan Unur

use strict;

my @variants = qw/
   agggtaaa|tttaccct
   [cgt]gggtaaa|tttaccc[acg]
   a[act]ggtaaa|tttacc[agt]t
   ag[act]gtaaa|tttac[agt]ct
   agg[act]taaa|ttta[agt]cct
   aggg[acg]aaa|ttt[cgt]ccct
   agggt[cgt]aa|tt[acg]accct
   agggta[cgt]a|t[acg]taccct
   agggtaa[cgt]|[acg]ttaccct
/;

my @variants_re = map qr/$_/xiaa, @variants;

my @iub = map { my $x = $_; sub { $_[0] =~ s/$x->[0]/$x->[1]/g }} (
    [ qr{ tHa [Nt] }x,                 ʼ<4>ʼ ],
    [ qr{ aND | caN | Ha[DS] | WaS }x, ʼ<3>ʼ ],
    [ qr{ a [NSt] | BY }x,             ʼ<2>ʼ ],
    [ qr{ < [^>]* > }x,                ʼ|ʼ   ],
    [ qr{ \| [^|] [^|]* \| }x,         ʼ-ʼ   ],
);

my $seq = do { local $/; <STDIN> };

my $input_length = length( $seq );

$seq =~ s/>.*\n|\n//g;

my $cleaned_length = length( $seq );

my @results = map scalar( () = $seq =~ /$_/g ), @variants;

$_->($seq) for @iub;

# report

print "$variants[$_] $results[$_]\n" for 0 .. $#variants;
print "$_\n" for ʼʼ, $input_length, $cleaned_length, length( $seq );


